### DISCUSSION TITLE
GETMONITORINFOEX struct not generated

### DISCUSSION THREAD

**Discussion by JesseAtExitNorth** [microsoft/CsWin32#188] [Q&A] [answered]
My NativeMethods.txt is simply `User32.*` and while the GETMONITORINFO struct is available (https://docs.microsoft.com/en-us/windows/win32/api/winuser/ns-winuser-monitorinfo) GETMONITORINFOEX wasn't generated.

I'm not sure if I am doing something incorrectly or this is a bug.

**Comment by AArnott** [comment, marked as answer]
There is no `GETMONITORINFO` struct, nor `GETMONITORINFOEX`. I think you meant `MONITORINFOEX` (no `GET` in front).
Structs aren't exported from User32 or any other module. They are simply in header files. We autogenerate structs when you ask for a method based on the method signature. You get `MONITORINFO` because that is used in the method signature, but `MONITORINFOEX` is not. 

To get `MONITORINFOEX`, you will need to ask for it explicitly in the NativeMethods.txt file.

**Reply by JesseAtExitNorth** [reply]
Yes, sorry about the "GET" typo. 

The NativeMethods.txt entry should just be `MONITORINFOEX` to generate that struct right? So 
`User32.*`
`MONITORINFOEX`
? I rebuilt the project but it didn't appear to generate MONITORINFOEX still.

**Reply by AArnott** [reply]
Ah, the reason for that is that `MONITORINFOEX` does not exist in the metadata. It varies between unicode and ansi versions so you have to request `MONITORINFOEXW` or `MONITORINFOEXA`.

**Reply by jnm2** [reply]
@AArnott Could the generator detect this kind of situation and provide a helpful error telling you what your options are?

**Reply by AArnott** [reply]
Filed an issue to track

**Reply by JesseAtExitNorth** [reply]
> Ah, the reason for that is that `MONITORINFOEX` does not exist in the metadata. It varies between unicode and ansi versions so you have to request `MONITORINFOEXW` or `MONITORINFOEXA`.

Thank you!

**Reply by dotMorten** [reply]
So if I add `MONITORINFOEXW` or `MONITORINFOEXA` to `NativeMethods.txt` those types do get generated, but I still can't call GetMonitorInfo though since it still generates with the `MONITORINFO` parameter which isn't a superclass of the EX types.

**Reply by AArnott** [reply]
`MONITORINFO*` types are structs, so they cannot be super-classes. Just like in C, you can allocate an EX struct, initialize its cbSize field, then pass a pointer to it to the `GetMonitorInfo` method, and it'll determine based on `cbSize` that you gave it the Ex struct and initialize its additional fields.
